packages/editor/src/AppModel/FieldModel.ts

```diff
@@ -109,28 +137,85 @@ export class FieldModel implements IFieldModel {
       parseExpression(this.value as string).filter(exp => typeof exp !== 'string')
     );
 
+    this.refs = {};
+    this.nodes = {};
+
     exps.forEach(exp => {
       let lastIdentifier: ComponentId = '' as ComponentId;
-      simpleWalk((acornLoose as typeof acorn).parse(exp, { ecmaVersion: 2020 }), {
-        Expression: node => {
-          switch (node.type) {
+      const node = (acornLoose as typeof acorn).parse(exp, { ecmaVersion: 2020 });
+
+      this.nodes[exp] = node;
+
+      simpleWalk(node, {
+        Expression: expressionNode => {
+          switch (expressionNode.type) {
             case 'Identifier':
-              const key = exp.slice(node.start, node.end) as ComponentId;
-              this.refs[key] = [];
+              const key = exp.slice(
+                expressionNode.start,
+                expressionNode.end
+              ) as ComponentId;
+
+              if (this.refs[key]) {
+                this.refs[key].nodes.push(expressionNode as Flatten<RefInfo['nodes']>);
+              } else {
+                this.refs[key] = {
+                  nodes: [expressionNode as Flatten<RefInfo['nodes']>],
+                  properties: [],
+                };
+              }
               lastIdentifier = key;
+
               break;
             case 'MemberExpression':
-              const str = exp.slice(node.start, node.end);
+              const str = exp.slice(expressionNode.start, expressionNode.end);
               let path = str.replace(lastIdentifier, '');
               if (path.startsWith('.')) {
                 path = path.slice(1, path.length);
               }
-              this.refs[lastIdentifier]?.push(path);
+              this.refs[lastIdentifier]?.properties.push(path);
               break;
             default:
           }
         },
       });
     });
   }
+
+  onReferenceIdChange({ oldId, newId }: { oldId: ComponentId; newId: ComponentId }) {
+    if (!this.componentModel) {
+      return;
+    }
+
+    if (this.spec?.isComponentId && this.value === oldId) {
+      if (this.traitModel) {
+        this.traitModel._isDirty = true;
+      }
+      this.componentModel._isDirty = true;
+      this.update(newId);
+    } else if (this.refs[oldId]) {
+      const exps = parseExpression(this.value as string);
+      const newExps = exps.map(exp => {
+        const node = this.nodes[exp.toString()];
+
+        if (node) {
+          const ref = this.refs[oldId];
+
+          ref.nodes.forEach(refNode => {
+            refNode.name = newId;
+          });
+
+          this.refs[newId] = ref;
+          delete this.refs[oldId];
+
+          return [escodegen.generate(node)];
+        }
+
+        return exp;
+      });
+      const value = expChunkToString(newExps);
+
+      this.componentModel._isDirty = true;
```

Comment:
Is is missing to set TraitModel isDirty here?

Reply:
Done.